glyphs2fontir/src/source.rs

```diff
@@ -258,20 +276,97 @@ impl Work for StaticMetadataWork {
     }
 }
 
+fn design_location(axes: &[ir::Axis], master: &FontMaster) -> DesignLocation {
+    axes.iter()
+        .zip(master.axes_values.as_ref().unwrap())
+        .map(|(axis, pos)| (axis.tag.clone(), DesignCoord::new(pos.into_inner() as f32)))
+        .collect()
+}
+
+fn to_normalized(axes: &[ir::Axis], design_location: &DesignLocation) -> NormalizedLocation {
+    let axes: HashMap<_, _> = axes.iter().map(|a| (&a.tag, a)).collect();
+    design_location
+        .iter()
+        .map(|(tag, dc)| {
+            (
+                tag.clone(),
+                dc.to_normalized(&axes.get(tag).unwrap().converter),
+            )
+        })
+        .collect()
+}
+
 struct GlyphIrWork {
     glyph_name: String,
+    font: Arc<Font>,
+    master_indices: Arc<HashMap<String, usize>>,
 }
 
 impl Work for GlyphIrWork {
     fn exec(&self, context: &Context) -> Result<(), WorkError> {
         debug!("Generate IR for {}", self.glyph_name);
+        let font = self.font.as_ref();
+
         let static_metadata = context.get_static_metadata();
+        let axes = &static_metadata.axes;
+
         let gid = static_metadata
             .glyph_id(&self.glyph_name)
             .ok_or_else(|| WorkError::NoGlyphIdForName(self.glyph_name.clone()))?;
+        let glyph = font
+            .glyphs
+            .get(&self.glyph_name)
+            .ok_or_else(|| WorkError::NoGlyphForName(self.glyph_name.clone()))?;
+
+        let mut sources = HashMap::new();
+
+        // Glyphs have layers that match up with masters, and masters have locations
+        let mut axis_positions: HashMap<String, HashSet<NormalizedCoord>> = HashMap::new();
+        for layer in glyph.layers.iter() {
+            let Some(master_idx) = self.master_indices.get(&layer.layer_id) else {
+                panic!("glyph {} references layer id {} but no such master exists", self.glyph_name, layer.layer_id);
+            };
+            let master = &font.font_master[*master_idx];
+            let location = to_normalized(axes, &design_location(axes, master));
```

Comment:
you are computing this location over and over for each glyph master layer, but it's always going to be the same for a given master. Maybe instead of a `master_indices` map {String: usize} you can store in the GlyphIrWork a `master_locations` hashmap keyed by master.id (string) and the values contain the precomputed normalized locations for each master.

Reply:
I miss Python's cache annotations

Reply:
We use axes from static metadata to compute this so I moved it outside the loop. It could be computed from the Font and shared globally but it makes the computation uglier.